actors/market/src/lib.rs

```diff
@@ -751,55 +752,34 @@ impl Actor {
                 let deal_ids = st.get_deals_for_epoch(rt.store(), i)?;
 
                 for deal_id in deal_ids {
-                    let deal = st.get_proposal(rt.store(), deal_id)?;
-                    let dcid = rt_deal_cid(rt, &deal)?;
-                    let state = st.find_deal_state(rt.store(), deal_id)?;
-
-                    // deal has been published but not activated yet -> terminate it
-                    // as it has timed out
-                    if state.is_none() {
-                        // Not yet appeared in proven sector; check for timeout.
-                        if curr_epoch < deal.start_epoch {
-                            return Err(actor_error!(
-                                illegal_state,
-                                "deal {} processed before start epoch {}",
-                                deal_id,
-                                deal.start_epoch
-                            ));
-                        }
-
-                        let slashed = st.process_deal_init_timed_out(rt.store(), &deal)?;
-                        if !slashed.is_zero() {
-                            amount_slashed += slashed;
+                    let deal_proposal = match st.get_proposal(rt.store(), deal_id) {
+                        Ok(dp) => dp,
+                        Err(err) => {
+                            if err.exit_code() == EX_DEAL_EXPIRED {
+                                // fine: deal may have expired and been deleted during synchronous PDU call
+                                continue;
+                            } else {
+                                return Err(err);
+                            }
                         }
+                    };
+                    let dcid = rt_deal_cid(rt, &deal_proposal)?;
 
-                        // Delete the proposal (but not state, which doesn't exist).
-                        let deleted = st.remove_proposal(rt.store(), deal_id)?;
-
-                        if deleted.is_none() {
-                            return Err(actor_error!(
-                                illegal_state,
-                                format!(
-                                    "failed to delete deal {} proposal {}: does not exist",
-                                    deal_id, dcid
-                                )
-                            ));
+                    let (state, maybe_slashed) = st.get_active_deal_or_cleanup(
```

Comment:
Could we name this `get_active_deal_or_expire`? And maybe_slashed to expiration_penalty.